src/provider/provider.go

```diff
@@ -0,0 +1,97 @@
+package provider
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+
+	logger "github.com/sirupsen/logrus"
+
+	"github.com/envoyproxy/ratelimit/src/config"
+)
+
+type Provider struct {
+	Endpoint     string
+	Subpath      []string
+	PollInterval time.Duration
+	PollTimeout  time.Duration
+	ConfigChan   chan []config.RateLimitConfigToLoad
+	httpClient   *http.Client
+	StopSig      bool
+	mu           sync.RWMutex
+}
+
+func NewProvider(endpoint string, subpath []string, pollInterval time.Duration, pollTimeout time.Duration) *Provider {
+	return &Provider{
+		Endpoint:     endpoint,
+		Subpath:      subpath,
+		PollInterval: pollInterval,
+		ConfigChan:   make(chan []config.RateLimitConfigToLoad),
+		httpClient: &http.Client{
+			Timeout: pollTimeout,
+		},
+		StopSig: false,
+	}
+}
+
+func (p *Provider) Provide() error {
+	sig := p.getStopSignal()
+	for !sig {
+		configs, err := p.fetchConfigurationData()
+		if err != nil {
+			logger.Errorf("Failed to fetch the configuration: %+v", err)
+			time.Sleep(p.PollInterval)
+			continue
+		}
+		p.ConfigChan <- configs
+		time.Sleep(p.PollInterval)
+	}
+	return nil
+}
+
+func (p *Provider) Stop() {
+	p.StopSig = true
+}
+
+func (p *Provider) getStopSignal() bool {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	sig := p.StopSig
+	return sig
+}
+
+// fetchConfigurationData fetches the configuration data from the configured endpoint.
+func (p *Provider) fetchConfigurationData() ([]config.RateLimitConfigToLoad, error) {
+	configs := []config.RateLimitConfigToLoad{}
+
+	urls := []string{}
+	if len(p.Subpath) == 0 {
+		urls = append(urls, p.Endpoint)
+	} else {
+		for _, path := range p.Subpath {
+			urls = append(urls, fmt.Sprintf("%s/%s", p.Endpoint, path))
+		}
+	}
+
+	for _, url := range urls {
+		res, err := p.httpClient.Get(url)
```

Comment:
Is it possible to add a timeout here?

Reply:
Yes, it is added when initializing the httpClient.